flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedLeaderService.java

```diff
@@ -244,38 +246,45 @@ private void removeContender(EmbeddedLeaderElection embeddedLeaderElection) {
     }
 
     /** Callback from leader contenders when they confirm a leader grant. */
+    @GuardedBy("lock")
     private void confirmLeader(
             final EmbeddedLeaderElection embeddedLeaderElection,
             final UUID leaderSessionId,
             final String leaderAddress) {
-        synchronized (lock) {
-            // if the leader election was shut down in the meantime, ignore this confirmation
-            if (!embeddedLeaderElection.running || shutdown) {
-                return;
-            }
-
-            try {
-                // check if the confirmation is for the same grant, or whether it is a stale grant
-                if (embeddedLeaderElection == currentLeaderProposed
-                        && currentLeaderSessionId.equals(leaderSessionId)) {
-                    LOG.info(
-                            "Received confirmation of leadership for leader {} , session={}",
-                            leaderAddress,
-                            leaderSessionId);
-
-                    // mark leadership
-                    currentLeaderConfirmed = embeddedLeaderElection;
-                    currentLeaderAddress = leaderAddress;
-                    currentLeaderProposed = null;
+        Preconditions.checkState(
+                currentLeaderProposed == embeddedLeaderElection,
+                "The confirmLeader method should only be called when having the leadership acquired.");
+        LOG.info(
+                "Received confirmation of leadership for leader {} , session={}",
+                leaderAddress,
+                leaderSessionId);
+
+        // mark leadership
+        currentLeaderConfirmed = embeddedLeaderElection;
+        currentLeaderAddress = leaderAddress;
+        currentLeaderProposed = null;
+
+        // notify all listeners
+        notifyAllListeners(leaderAddress, leaderSessionId);
+    }
 
-                    // notify all listeners
-                    notifyAllListeners(leaderAddress, leaderSessionId);
-                } else {
-                    LOG.debug(
-                            "Received confirmation of leadership for a stale leadership grant. Ignoring.");
+    private void runAsLeader(
+            EmbeddedLeaderElection embeddedLeaderElection,
+            UUID leaderSessionId,
+            ThrowingRunnable<? extends Throwable> runnable)
+            throws LeadershipLostException {
+        synchronized (lock) {
+            if (embeddedLeaderElection.running
+                    && !shutdown
+                    && embeddedLeaderElection.isLeader
+                    && currentLeaderSessionId.equals(leaderSessionId)) {
+                try {
+                    runnable.run();
+                } catch (Throwable t) {
+                    fatalError(t);
                 }
-            } catch (Throwable t) {
-                fatalError(t);
+            } else {
+                throw new LeadershipLostException(leaderSessionId);
```

Comment:
I'm not quite sold on this. It feels like we're using a special exception to do control flow (such that users can ignore certain exceptions).

Reply:
For me, it was meant as an invalid state: The callback that's passed into the `runAsLeader` method should succeed in the happy scenario. Losing the leadership in the mean time is a an "unusual" state of the system. In this sense, I felt like exceptions being the most appropriate tool to express this.

You're suggesting that we should make `runAsLeader` return `Completable<Boolean>` indicating whether the callback was executed or not? That would work as well. But from an interface perspective, `runAsLeader` is just one way to execute the callback. One could also (theoretically) think of a `supplyAsLeader` (as it's implemented in `JobMasterServiceLeadershipRunner` right now). In such a case, utilizing the return value of `CompletableFuture` and relying on an exception for leadership loss feels like the more natural thing to do.